core/types/permissions_cache.go

```diff
@@ -0,0 +1,410 @@
+	c, _ := lru.New(defaultMapLimit)
+	return &OrgCache{c, sync.Mutex{}}
+}
+
+func NewNodeCache() *NodeCache {
+	c, _ := lru.New(defaultMapLimit)
+	return &NodeCache{c, sync.Mutex{}}
+}
+
+func NewRoleCache() *RoleCache {
+	c, _ := lru.New(defaultMapLimit)
+	return &RoleCache{c, sync.Mutex{}}
+}
+
+func NewAcctCache() *AcctCache {
+	c, _ := lru.New(defaultMapLimit)
+	return &AcctCache{c, sync.Mutex{}}
+}
+
+var syncStarted = false
+
+var DefaultAccess = FullAccess
+var networkAdminRole string
+var orgAdminRole string
+
+const defaultMapLimit = 100
+
+//var OrgKeyMap, _ = lru.New(orgKeyMapLimit)
+
+var OrgInfoMap = NewOrgCache()
+var NodeInfoMap = NewNodeCache()
+var RoleInfoMap = NewRoleCache()
+var AcctInfoMap = NewAcctCache()
+
+func (pc *PermissionConfig) IsEmpty() bool {
+	return pc.InterfAddress == common.HexToAddress("0x0") || pc.NodeAddress == common.HexToAddress("0x0") || pc.AccountAddress == common.HexToAddress("0x0")
+}
+
+func SetSyncStatus() {
+	syncStarted = true
+}
+
+func GetSyncStatus() bool {
+	return syncStarted
+}
+
+// sets default access to readonly and initializes the values for
+// network admin role and org admin role
+func SetDefaults(nwRoleId, oaRoleId string) {
+	DefaultAccess = ReadOnly
+	networkAdminRole = nwRoleId
+	orgAdminRole = oaRoleId
+}
+
+func (o *OrgCache) UpsertOrg(orgId, parentOrg, ultimateParent string, level *big.Int, status OrgStatus) {
+	defer o.mux.Unlock()
+	o.mux.Lock()
+	var key OrgKey
+	if parentOrg == "" {
+		key = OrgKey{OrgId: orgId}
+	} else {
+		key = OrgKey{OrgId: parentOrg + "." + orgId}
+		pkey := OrgKey{OrgId: parentOrg}
+		if ent, ok := o.c.Get(pkey); ok {
+			porg := ent.(*OrgInfo)
+			if !containsKey(porg.SubOrgList, key.OrgId) {
+				porg.SubOrgList = append(porg.SubOrgList, key.OrgId)
+				o.c.Add(pkey, porg)
+			}
+		}
+	}
+
+	norg := &OrgInfo{orgId, key.OrgId, parentOrg, ultimateParent, level, nil, status}
+	o.c.Add(key, norg)
+}
+
+func containsKey(s []string, e string) bool {
+	for _, a := range s {
+		if a == e {
+			return true
+		}
+	}
+	return false
+}
+
+func (o *OrgCache) GetOrg(orgId string) *OrgInfo {
+	defer o.mux.Unlock()
+	o.mux.Lock()
+	key := OrgKey{OrgId: orgId}
+	if ent, ok := o.c.Get(key); ok {
+		return ent.(*OrgInfo)
+	}
+	return nil
+}
+
+func (o *OrgCache) GetOrgList() []OrgInfo {
+	var olist []OrgInfo
+	for _, k := range o.c.Keys() {
+		v, _ := o.c.Get(k)
+		vp := v.(*OrgInfo)
+		olist = append(olist, *vp)
+	}
+	return olist
+}
+
+func (n *NodeCache) UpsertNode(orgId string, url string, status NodeStatus) {
+	defer n.mux.Unlock()
+	n.mux.Lock()
+	key := NodeKey{OrgId: orgId, Url: url}
+	n.c.Add(key, &NodeInfo{orgId, url, status})
+}
+
+func (n *NodeCache) GetNodeByUrl(url string) *NodeInfo {
+	defer n.mux.Unlock()
+	n.mux.Lock()
+	for _, k := range n.c.Keys() {
+		ent := k.(NodeKey)
+		if ent.Url == url {
+			v, _ := n.c.Get(ent)
+			return v.(*NodeInfo)
+		}
+	}
+	return nil
+}
+
+func (o *NodeCache) GetNodeList() []NodeInfo {
+	var olist []NodeInfo
```

Comment:
if we know the length of `[]NodeInfo` in advanced, it'd be better to initialize the exact slice length

Reply:
handled for all the lists